### PR TITLE
[ODS-4443] Convert Specflow tests in EdFi.Ods.WebService.Tests to .NET Core for Composites

### PR DESCRIPTION
**Pull request by jamessmckay**

### CODE REVIEW

File: tests/EdFi.Ods.WebApi.IntegrationTests/UriHelper.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
+﻿// SPDX-License-Identifier: Apache-2.0
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+// See the LICENSE and NOTICES files in the project root for more information.
+
+namespace EdFi.Ods.WebApi.IntegrationTests
+{
+    internal class UriHelper
```
Comment (blmeyers) [inline tests/EdFi.Ods.WebApi.IntegrationTests/UriHelper.cs:8]: We have this duplicated now between the two projects, should the url constants move into Test.Common?

**Review by blmeyers — CHANGES_REQUESTED** [review]

File: Application/EdFi.Ods.Api/Helpers/TypeHelper.cs [hunk 2]
```diff
@@ -19,22 +21,29 @@ public static class TypeHelper
...
             assemblies.ForEach(a => _logger.Debug($"{a.GetName().Name}"));
 
-            return assemblies
-                .SelectMany(
-                    a => a.GetTypes()
-                        .Where(
-                            t => t.GetInterfaces()
-                                .Contains(typeof(IModule)) && !t.IsAbstract));
+            // force all test modules last
+            var overridingModules = assemblies
```
[the remark this answers is not in the document]
Comment (blmeyers) [reply Application/EdFi.Ods.Api/Helpers/TypeHelper.cs:31]: https://tracker.ed-fi.org/browse/ODS-4590

[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.Api/Helpers/TypeHelper.cs:31]: done, removed the "Tests" dependency on shared code, and renamed the test module to use override

File: Application/EdFi.Ods.Repositories.NHibernate.Tests/EdFi.Ods.Repositories.NHibernate.Tests.csproj [hunk 3]
```diff
@@ -9,6 +9,7 @@
     <OutputType>Library</OutputType>
     <IsTestProject>true</IsTestProject>
     <NoWarn>CS0618</NoWarn>
+    <NoDefaultLaunchSettingsFile>true</NoDefaultLaunchSettingsFile>
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.Repositories.NHibernate.Tests/EdFi.Ods.Repositories.NHibernate.Tests.csproj:12]: I found this by accident. I brought up the solution in Visual Studio and it created launch settings for the test projects, however we do not need them, so this fixes that problem. Launch settings are created when the project definition using `<Project Sdk="Microsoft.NET.Sdk.Web">`

File: Application/EdFi.Ods.WebService.Tests/EdFi.Ods.WebService.Tests.csproj [hunk 4]
```diff
@@ -47,10 +47,7 @@
     <PackageReference Include="SpecFlow.NUnit" Version="3.3.74" />
     <PackageReference Include="SpecFlow.Tools.MsBuild.Generation" Version="3.3.74" />
   </ItemGroup>
-  <ItemGroup Condition=" '$(TargetFramework)' == 'netcoreapp3.1' ">
-    <PackageReference Include="Autofac.Extensions.DependencyInjection" Version="6.0.0" />
-  </ItemGroup>
-  <ItemGroup Condition=" '$(TargetFramework)' == 'net48' ">
+    <ItemGroup Condition=" '$(TargetFramework)' == 'net48' ">
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.WebService.Tests/EdFi.Ods.WebService.Tests.csproj:50]: done

File: Application/EdFi.Ods.WebService.Tests/EdFi.Ods.WebService.Tests.csproj [hunk 5]
```diff
@@ -1,6 +1,6 @@
-﻿<Project Sdk="Microsoft.NET.Sdk.Web">
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Application/EdFi.Ods.WebService.Tests/EdFi.Ods.WebService.Tests.csproj:1]: correct

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/AssessmentKeyInformationDto.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: Apache-2.0
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+// See the LICENSE and NOTICES files in the project root for more information.
+
+using System;
+
+namespace EdFi.Ods.WebApi.CompositeSpecFlowTests
+{
+    public class AssessmentKeyInformationDto
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/AssessmentKeyInformationDto.cs:10]: removed the suffix

[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/AssessmentKeyInformationDto.cs:10]: also moved to a dtos namespace, and renamed the objects to be more meaningful

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/CompositesSpecFlowTestFixture.cs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+                            webBuilder.UseStartup<Startup>();
+                            webBuilder.UseUrls(CompositesTestConstants.BaseUrl);
+                        })
+                    .Build();
+
+                ServiceProvider = Host.Services;
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine(e);
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/CompositesSpecFlowTestFixture.cs:59]: removed

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/EdFi.Ods.WebApi.CompositeSpecFlowTests.csproj [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    <Content Update="appsettings.json">
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </Content>
+    <Content Update="log4net.config">
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </Content>
+  </ItemGroup>
+  <ItemGroup>
+    <None Update="hibernate.cfg.xml">
+      <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/EdFi.Ods.WebApi.CompositeSpecFlowTests.csproj:68]: removed

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/FakedCompositesOdsConnectionStringProvider.cs [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+// See the LICENSE and NOTICES files in the project root for more information.
+
+using EdFi.Ods.Common.Database;
+
+namespace EdFi.Ods.WebApi.CompositeSpecFlowTests
+{
+    public class FakedCompositesOdsConnectionStringProvider : IOdsDatabaseConnectionStringProvider
+    {
+        public string GetConnectionString()
+            => $"Server=(local); Database={CompositesSpecFlowTestFixture.SpecFlowDatabaseName}; Trusted_Connection=True; Application Name=EdFi.Ods.WebApi;";
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/FakedCompositesOdsConnectionStringProvider.cs:13]: changed to
```
        public async Task ShouldBuildDbAndSetupWebServer()
        {
            var connectionStringProvider = (IOdsDatabaseConnectionStringProvider)
                CompositesSpecFlowTestFixture.ServiceProvider.GetService(typeof(IOdsDatabaseConnectionStringProvider));
            
            connectionStringProvider.ShouldNotBeNull();
            connectionStringProvider.GetConnectionString().ShouldContain(CompositesSpecFlowTestFixture.SpecFlowDatabaseName);
            
            var cancellationToken = new CancellationToken();
            await using var conn = new SqlConnection(connectionStringProvider.GetConnectionString());

            await conn.OpenAsync(cancellationToken);

            int count = await conn.QuerySingleAsync<int>("select count(*) from dbo.DeployJournal;", cancellationToken);

            count.ShouldNotBe(0);

            var response = await new HttpClient().GetAsync(CompositesTestConstants.BaseUrl, cancellationToken);

            response.StatusCode.ShouldBe(HttpStatusCode.OK);

            var json = await response.Content.ReadAsStringAsync();

            json.ShouldNotBeNullOrWhiteSpace();
        }
```

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/KeyInformationDto.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
+// SPDX-License-Identifier: Apache-2.0
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+// See the LICENSE and NOTICES files in the project root for more information.
+
+using System;
+
+namespace EdFi.Ods.WebApi.CompositeSpecFlowTests
+{
+    public class KeyInformationDto
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/KeyInformationDto.cs:10]: done

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/CompositesSpecFlowTestFixture.cs [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+                throw;
+            }
+
+            var configuration = (IConfigurationRoot) ServiceProvider.GetService(typeof(IConfiguration));
+
+            var PopulatedTemplateDatabaseName = configuration.GetSection("TestDatabaseTemplateName").Value;
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/CompositesSpecFlowTestFixture.cs:65]: done

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/EnvironmentSetupTests.cs [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace EdFi.Ods.WebApi.CompositeSpecFlowTests
+{
+    [TestFixture]
+    public class EnvironmentSetupTests
+    {
+        [Test]
+        public async Task ShouldBuildDbAndSetupWebServer()
+        {
+            string connectionString =
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/EnvironmentSetupTests.cs:23]: done

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/StepsHelper.cs [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            await using var conn = new SqlConnection(connectionString);
+
+            await conn.OpenAsync(cancellationToken);
+            return await conn.QuerySingleOrDefaultAsync<T>(query, cancellationToken);
+        }
+
+        public static string GetQueryString(string correlationId)
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/StepsHelper.cs:46]: done

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/StepsHelper.cs [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+            await conn.OpenAsync(cancellationToken);
+
+            string query = $@"
+                  SELECT Id
+                  FROM  [edfi].[{tableName}]
+                  WHERE {whereClause}";
+
+            return await conn.QuerySingleOrDefaultAsync<Guid>(query, cancellationToken);
+
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/StepsHelper.cs:85]: done

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs [hunk 15]
```diff
@@ -0,0 +1,570 @@
...
+
+            await SetIdAsync(query);
+            SetResource("StudentEducationOrganizationAssociation");
+        }
+
+        [Given(@"the subject of the request is a school with student and staff associations")]
+        public async Task GivenTheSubjectOfTheRequestIsASchoolWithStudentAndStaffAssociations()
+        {
+            const string query = @"
+                    SELECT  TOP 1 eo.Id, *
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs:67]: fixed

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs [hunk 16]
```diff
@@ -0,0 +1,570 @@
...
+
+            await SetIdAsync(query);
+            SetResource("School");
+        }
+
+        [Given(@"the subject of the request is a student with a StudentAcademicRecord")]
+        public async Task GivenTheSubjectOfTheRequestIsAStudentWithAStudentAcademicRecord()
+        {
+            const string query = @"
+                    SELECT  TOP 1 s.Id, *
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs:82]: done

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs [hunk 17]
```diff
@@ -0,0 +1,570 @@
...
+
+            await SetIdAsync(query);
+            SetResource("Student");
+        }
+
+        [Given(@"the subject of the request is a student with values in all name properties")]
+        public async Task GivenTheSubjectOfTheRequestIsAStudentWithValuesInAllNameProperties()
+        {
+            const string query = @"
+                    SELECT  TOP 1 Id, *
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs:94]: done

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs [hunk 18]
```diff
@@ -0,0 +1,570 @@
...
+
+            await SetIdAsync(query);
+            SetResource("StudentSchoolAssociation");
+        }
+
+        [Given(@"the subject of the request is a StudentSchoolAssociation with StudentAssessment")]
+        public async Task GivenTheSubjectOfTheRequestIsAStudentSchoolAssociationWithStudentAssessment()
+        {
+            const string query = @"
+                    SELECT  TOP 1 ssa.Id, stu.StudentUniqueId, ssa.SchoolId, ssa.EntryDate
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs:121]: done

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/KeyInformationDto.cs [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        public Guid Id { get; set; }
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/KeyInformationDto.cs:12]: done

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs [hunk 20]
```diff
@@ -0,0 +1,570 @@
...
+
+            await SetKeyInformationAsync(query);
+            SetResource("StudentSchoolAssociation");
+        }
+
+        [Given(@"the subject of the request is a StudentSchoolAssociation with School")]
+        public async Task GivenTheSubjectOfTheRequestIsAStudentSchoolAssociationWithSchool()
+        {
+            const string query = @"
+                    SELECT  TOP 1 ssa.Id, stu.StudentUniqueId, ssa.SchoolId, ssa.EntryDate
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs:136]: done

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs [hunk 21]
```diff
@@ -0,0 +1,570 @@
...
+
+            await SetIdAsync(query);
+            SetResource("StudentAssessment");
+        }
+
+        [Given(@"the subject of the request is a StudentAssessment with StudentAssessmentStudentObjectiveAssessment")]
+        public async Task GivenTheSubjectOfTheRequestIsAStudentAssessmentWithStudentAssessmentStudentObjectiveAssessment()
+        {
+            const string query = @"
+                    SELECT  TOP 1 sa.Id,
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs:168]: done

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs [hunk 22]
```diff
@@ -0,0 +1,570 @@
...
+
+            await SetAssessmentKeyInformationAsync(query);
+            SetResource("StudentAssessment");
+        }
+
+        [Given(@"the subject of the request is a student with a StudentSchoolAssociation")]
+        public async Task GivenTheSubjectOfTheRequestIsAStudentWithAStudentSchoolAssociation()
+        {
+            const string query = @"
+                    SELECT  TOP 1 s.Id, *
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs:190]: done

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs [hunk 23]
```diff
@@ -0,0 +1,570 @@
...
+
+            await SetIdAsync(query);
+            SetResource("Student");
+        }
+
+        [Given(@"the subject of the request is a student with a StudentEducationOrganizationAssociation")]
+        public async Task GivenTheSubjectOfTheRequestIsAStudentWithAStudentEducationOrganizationAssociation()
+        {
+            const string query = @"
+                    SELECT  TOP 1 s.Id, *
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs:202]: done

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs [hunk 24]
```diff
@@ -0,0 +1,570 @@
...
+
+            await SetIdAsync(query);
+            SetResource("Student");
+        }
+
+        [Given(@"the subject of the request is a StudentSchoolAssociation with School and StudentAssessment")]
+        public async Task GivenTheSubjectOfTheRequestIsAStudentSchoolAssociationWithSchoolAndStudentAssessment()
+        {
+            const string query = @"
+                    SELECT  TOP 1 ssa.Id, stu.StudentUniqueId, ssa.SchoolId, ssa.EntryDate
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs:214]: done

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs [hunk 25]
```diff
@@ -0,0 +1,570 @@
...
+
+                    throw new NotSupportedException(
+                        string.Format("No request definition matched '{0}'.", requestName));
+            }
+
+            string correlationId = Guid.NewGuid().ToString("n");
+
+            SetCorrelationId(correlationId);
+
+            requestUrl += string.Format("?{0}={1}", SpecialQueryStringParameters.CorrelationId, correlationId);
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs:458]: done

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs [hunk 26]
```diff
@@ -0,0 +1,570 @@
...
+                var httpClient = StepsHelper.GetHttpClient();
+                var response = await httpClient.GetAsync(uri, _cancellationToken.Value);
+
+                response.StatusCode.ShouldBe(HttpStatusCode.OK);
+
+                json = await response.Content.ReadAsStringAsync();
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine(e);
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs:475]: done

File: tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs [hunk 27]
```diff
@@ -0,0 +1,570 @@
...
+            {
+                {"StudentUniqueId", dto.StudentUniqueId},
+                {"SchoolId", dto.SchoolId},
+                {"EntryDate", dto.EntryDate}
+            };
+
+            _scenarioContext.Set(keyValueByName, ScenarioContextKeys.CompositeSubjectKey);
+        }
+
+        private async Task SetAssessmentKeyInformationAsync(string query)
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.CompositeSpecFlowTests/Steps.cs:541]: done

File: tests/EdFi.Ods.WebApi.IntegrationTests/UriHelper.cs (hunk 1 above)
Comment (jamessmckay) [reply tests/EdFi.Ods.WebApi.IntegrationTests/UriHelper.cs:8]: Yes, i will make it a class we can new and pass in the base address

**Review by tbmeador-dlp — APPROVED** [review]
approving commit 3d0cfa59e417e055f12c3b2c8bb2861bf6038503

**Review by blmeyers — APPROVED** [review]

**Pull request merged**